design/service-apis-implementation.md

```diff
@@ -0,0 +1,229 @@
+
+The Service APIs are a set of interrelated Kubernetes objects, where a change in one object can mean that the scope of objects Contour is interested in will change.
+Because of this, Contour will watch all events associated with GatewayClass, Gateway, HTTPRoute, TLSRoute, and BackendPolicy objects, and filter the objects it takes action on internally, rather than using a filtered watch (as those are costly to set up and tear down).
+
+### Combining Service APIs with other configuration
+
+When it calculates the Envoy configuration using the DAG, Contour layers different types of configuration in order.
+So, currently, Ingress is overwritten by HTTPProxy, so if an Ingress and a HTTPProxy specify the exact same route, the HTTPProxy will win.
+
+Once we have the Service-APIs available as well, we have to choose the order.
+In this design, I suggest having the order be Ingress is overwritten by HTTPProxy, is overwritten by the service-apis.
+I could see reversing HTTPProxy and the Service APIs here, but I think this acknowledges that the Service APIs are really an evolution of the ideas in HTTPProxy, and will probably end up being the community standard.
+
+### Status management
+
+Each object in the Service APIs set has its own status definition. In general, Contour will update the status of any object that comes into scope with its details, ensuring that the `observedGeneration` field is set correctly.
+When objects fall out of scope, any status set by Contour will not be removed.
+It's expected that things that check the status will also check the `observedGeneration` to check if the status information is up-to-date with the current object generation.
+
+### Per-object design notes
+
+#### GatewayClass
+
+Contour will look for entries with `spec.controller` equal to `projectcontour.io/contour` by default.
+Contour can change this value to any `projectcontour.io/<value>` with a config setting (similar to how `ingress.class` works today.)
+Importantly, this means that changing this values requires a Contour restart; this means that the GatewayClass watcher can have this value baked in at runtime.
+
+If Contour finds more than one matching GatewayClass, this is a conflict, and Contour will use the oldest GatewayClass (by creation timestamp).
+All GatewayClasses will be updated with status conditions indicating if they are in use.
+
+The output of this watcher is a list of GatewayClass name/namespace details.
+
+#### Gateway
+Contour will watch for entries with known GatewayClass name/namespace details, from the previous watch.
+
+Contour is not able to specify the address that the Envoy deployment will listen on, and so having any entries in Gateway's `spec.addresses` stanza will result in Contour marking those addresses as invalid in the Gateway's status.
+Contour expects the `spec.addresses` field to be empty.
+
+For Listeners:
+- Listeners inside Gateways will be merged where possible, as will listeners across Gateways.
```

Comment:
Contour only supports two listeners today (e.g. http/https). Does this statement assume that you are permitted to create more dynamically?

Reply:
The insecure/secure split definitely needs some thinking as a result of this. The initial  thought is that we would replicate the current behavior, which would mean accepting only one http and one https listener, but this will be tricky to do. We may need to do something different with automatic http->https redirect for the Service APIs.

Reply:
I've opened #3263 with a new design for this, and will update this document accordingly.